olp-cpp-sdk-core/src/network2/winhttp/NetworkProtocolWinHttp.cpp

```diff
@@ -0,0 +1,1063 @@
+    return NULL;
+  }
+  LPWSTR buffer = (LPWSTR)LocalAlloc(LPTR, len);
+  if (!buffer) return NULL;
+  index = WINHTTP_NO_HEADER_INDEX;
+  if (!WinHttpQueryHeaders(request, header, WINHTTP_HEADER_NAME_BY_INDEX,
+                           buffer, &len, &index)) {
+    LocalFree(buffer);
+    return NULL;
+  }
+  return buffer;
+}
+
+void UnixTimeToFileTime(std::uint64_t t, LPFILETIME pft) {
+  // Note that LONGLONG is a 64-bit value
+  LONGLONG ll;
+
+  ll = Int32x32To64(t, 10000000) + 116444736000000000;
+  pft->dwLowDateTime = (DWORD)ll;
+  pft->dwHighDateTime = ll >> 32;
+}
+
+bool convertMultiByteToWideChar(const std::string& s_in, std::wstring& s_out) {
+  s_out.clear();
+  if (s_in.empty()) {
+    return true;
+  }
+
+  // Detect required buffer size.
+  const auto chars_required = MultiByteToWideChar(
+      CP_ACP, MB_PRECOMPOSED, s_in.c_str(),
+      -1,       // denotes null-terminated string
+      nullptr,  // output buffer is null means request required buffer size
+      0);
+
+  if (chars_required == 0) {
+    // error: could not convert input string
+    return false;
+  }
+
+  s_out.resize(chars_required);
+  // Perform actual conversion from multi-byte to wide char
+  const auto conversion_result =
+      MultiByteToWideChar(CP_ACP, MB_PRECOMPOSED, s_in.c_str(),
+                          -1,  // denotes null-terminated string
+                          &s_out.front(), s_out.size());
+
+  if (conversion_result == 0) {
+    // Should not happen as 1st call have succeeded.
+    return false;
+  }
+  return true;
+}
+}  // namespace
+
+namespace olp {
+namespace network2 {
+
+#define LOGTAG "WinHttp"
+
+NetworkProtocolWinHttp::NetworkProtocolWinHttp()
+    : http_session_(NULL),
+      thread_(INVALID_HANDLE_VALUE),
+      event_(INVALID_HANDLE_VALUE) {
+  recent_request_id_.store(1);
+  http_session_ = WinHttpOpen(L"NGMOS CLient", WINHTTP_ACCESS_TYPE_NO_PROXY,
+                              WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS,
+                              WINHTTP_FLAG_ASYNC);
+
+  if (!http_session_) {
+    EDGE_SDK_LOG_ERROR(LOGTAG, "WinHttpOpen failed " << GetLastError());
+    return;
+  }
+
+  WinHttpSetStatusCallback(
+      http_session_,
+      (WINHTTP_STATUS_CALLBACK)&NetworkProtocolWinHttp::RequestCallback,
+      WINHTTP_CALLBACK_FLAG_ALL_COMPLETIONS | WINHTTP_CALLBACK_FLAG_HANDLES, 0);
+
+  event_ = CreateEvent(NULL, TRUE, FALSE, NULL);
+
+  // Store the memory tracking state during initialization so that it can be
+  // used by the thread.
+  thread_ = CreateThread(NULL, 0, NetworkProtocolWinHttp::Run, this, 0, NULL);
+  SetThreadPriority(thread_, THREAD_PRIORITY_ABOVE_NORMAL);
+}
+
+NetworkProtocolWinHttp::~NetworkProtocolWinHttp() {
+  std::vector<std::shared_ptr<ResultData>> pendingResults;
+  {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    std::vector<std::shared_ptr<RequestData>> requestsToCancel;
+    for (auto& request : http_requests_) {
```

Comment:
why do we keep the map around anymore? You can just as easy move the map and use it instead of copying to a vector

```cpp
{
    std::unique_lock<std::recursive_mutex> lock(mutex_);
    auto pending_requests = std::move(http_requests_);

    for (auto& request : pending_requests) {
      ....
    }
}
```